### DISCUSSION TITLE
Request Enhancement To Magick Compare -metric MSE

### DISCUSSION THREAD

**Discussion by fmw42** [ImageMagick/ImageMagick#5750] [Development]
UrbanWarrior has implemented my 100 fold speed up to **magick compare -metric ncc** using my FFT processing script as a guide. I would like to offer my **-metric MSE** script and request that it be implemented also to improve performance of that metric.  Code should be rather equivalent to what is developed now for -metric NCC.

**Comment by urban-warrior** [comment]
Post a link to your script with sample images and expected results.

**Comment by fmw42** [comment]
Here is a folder that has 1) the mse script and for comparison the ncc script, 2) two sets of examples.  Note that since the MSE is best at zero, I inverted the correlation surface image so that the best spot is brightest. I also included a -auto-level of the correlation surface for better viewing.

Here is a copy of the examples that are included in the folder:

```
Example 1:

bash fft_mse hat_124_124_417_44_white.png logo.png logo_corr.png
Match Coords: (417,44) And Score In Range 0 to 1: (0). <---- perfect match at 0

magick logo_corr.png -auto-level logo_corr_autolevel.png


Example 2:

bash fft_mse zelda3_32_27.png zelda3.png zelda3_corr.png
Match Coords: (32,27) And Score In Range 0 to 1: (7.15256e-07)  <---- near perfect match

magick zelda3_corr.png -auto-level zelda3_corr_autolevel.png

```

All the scripts and input and output images are in

[compare_mse.zip](https://github.com/ImageMagick/ImageMagick/files/9996408/compare_mse.zip)

**Revisions:**

[compare_mse_updates.zip](https://github.com/user-attachments/files/18241740/compare_mse_updates.zip)


Note that my script makes two output images that are not the same as what are in compare. Those are the input with a box around the match area and a crop of the best location in the input.  These are not necessary for compare.

**Comment by urban-warrior** [comment]
We added support for accelerated MSE similarity search as you requested.  Your script and algorithm documentation was quite helpful.

**Comment by fmw42** [comment]
Cristy,

Thanks.  If this works, you can easily add RMSE and PSNR. They differ from MSE by only the compare metric equation.  That is you use the same computations as MSE and just change the final computation after computing up to the MSE metric. I will send you the formulae after checking MSE. 

Fred

**Reply by fmw42** [reply]
NCC, MSE, RMSE, PSNR and new DPC and PHASE have all been added now as of IM 7.1.2.0. See https://github.com/ImageMagick/ImageMagick/discussions/3950
